File: 2781.length-of-the-longest-valid-substring.py
#
# @lc app=leetcode id=2781 lang=python3
#
# [2781] Length of the Longest Valid Substring
#

# @lc code=start
from typing import List
class Solution:
    def longestValidSubstring(self, word: str, forbidden: List[str]) -> int:
        # substring of word is valid if it does not contain any forbidden string
        # find the longest valid substring
        
        #remove all the duplicated forbidden strings
        forbidden = set(forbidden)
        ans = left = 0
        for right in range(len(word)):
            for i in range(right, max(right - 10, left -1), -1):
                if word[i: right + 1] in forbidden:
                    left = i + 1
                    break
            ans = max(ans, right - left + 1)
        return ans
        
# @lc code=end

File: 2781.length-of-the-longest-valid-substring.java
/*
 * @lc app=leetcode id=2781 lang=java
 *
 * [2781] Length of the Longest Valid Substring
 */
import java.util.*;
// @lc code=start
class Solution {
    public int longestValidSubstring(String word, List<String> forbidden) {
        Map<String, Integer> forbiddenMap = new HashMap<>();
        for (String s : forbidden) {
            forbiddenMap.put(s, 1);
        }
        int left = 0;
        int ans = 0;
        int n = word.length();
        for(int i = 0; i < n; i++){
            for(int j = i; j >= left && j > i - 10; j--){
                if(forbiddenMap.containsKey(word.substring(j, i + 1))){
                    left = j + 1;
                }
            }
            ans = Math.max(ans, i - left + 1);
        }
        return ans;
    }
}
// @lc code=end

